server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -476,17 +490,17 @@ private static void addSortedBundle(
         sortedBundles.add(bundle);
     }
 
-    private List<Tuple<PluginInfo, Plugin>> loadBundles(Set<Bundle> bundles) {
-        List<Tuple<PluginInfo, Plugin>> plugins = new ArrayList<>();
-        Map<String, Tuple<Plugin, ClassLoader>> loaded = new HashMap<>();
+    private List<LoadedPlugin> loadBundles(Set<Bundle> bundles) {
+        List<LoadedPlugin> plugins = new ArrayList<>();
+        Map<String, PluginClassloaderContext> loaded = new HashMap<>();
         Map<String, Set<URL>> transitiveUrls = new HashMap<>();
         List<Bundle> sortedBundles = sortBundles(bundles);
         for (Bundle bundle : sortedBundles) {
             if (bundle.plugin.getType() != PluginType.BOOTSTRAP) {
                 checkBundleJarHell(JarHell.parseClassPath(), bundle, transitiveUrls);
 
                 final Plugin plugin = loadBundle(bundle, loaded);
-                plugins.add(new Tuple<>(bundle.plugin, plugin));
+                plugins.add(new LoadedPlugin(bundle.plugin, plugin));
```

Comment:
Instead of adding these plugins as we go, we can get the `values()` from `loaded` at the end of this method.